src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/deciders/HeapBasedDecider.java

```diff
@@ -0,0 +1,77 @@
+      .put(UsageBucket.UNDER_UTILIZED, 10.0)
+      .put(UsageBucket.HEALTHY_WITH_BUFFER, 60.0)
+      .put(UsageBucket.HEALTHY, 80.0)
+      .build();
+
+  private HighHeapUsageClusterRca highHeapUsageClusterRca;
+
+  public HeapBasedDecider(long evalIntervalSeconds, int decisionFrequency, HighHeapUsageClusterRca highHeapUsageClusterRca) {
+    super(evalIntervalSeconds, decisionFrequency);
+    this.highHeapUsageClusterRca = highHeapUsageClusterRca;
+  }
+
+  /**
+   * The Queue and Cache deciders should only be able to suggest increase of the queue size or increase of cache size if the Java heap can
+   * sustain more live objects in it without de-gradation. What is an acceptable heap usage limit to determine this, comes from the
+   * bucketization object in rca.conf. We compare the oldGen usage percent reported by the HighHeapUsage RCA to determine that.
+   *
+   * @param esNode The EsNode we are trying to make a decision for.
+   * @return return if the OldGen heap is under-utilized or healthy and yet more can be consumed, return true; or false otherwise.
+   */
+  protected boolean canUseMoreHeap(NodeKey esNode) {
+    // we add action only if heap is under-utilized or healthy and yet more can be consumed.
+    for (ResourceFlowUnit<HotClusterSummary> clusterSummary : highHeapUsageClusterRca.getFlowUnits()) {
+      if (clusterSummary.hasResourceSummary()) {
+        for (HotNodeSummary nodeSummary : clusterSummary.getSummary().getHotNodeSummaryList()) {
+          NodeKey thisNode = new NodeKey(nodeSummary.getNodeID(), nodeSummary.getHostAddress());
+          if (thisNode.equals(esNode)) {
+            for (HotResourceSummary hotResourceSummary : nodeSummary.getHotResourceSummaryList()) {
+              Resource resource = hotResourceSummary.getResource();
+              if (resource.getResourceEnum() == DECIDING_HEAP_RESOURCE_TYPE) {
+                double oldGenUsedRatio = hotResourceSummary.getValue();
+                double oldGenUsedPercent = oldGenUsedRatio * 100;
+                BucketCalculator bucketCalculator;
+                try {
+                  bucketCalculator = rcaConf.getBucketizationSettings(OLD_GEN_TUNABLE_KEY);
+                } catch (Exception jsonEx) {
+                  bucketCalculator = new BasicBucketCalculator(DEFAULT_HEAP_USAGE_THRESHOLDS);
+                  LOG.debug("rca.conf does not have bucketization limits specified. Using default map.");
+                }
+                UsageBucket bucket = bucketCalculator.compute(oldGenUsedPercent);
+                LOG.debug("Value ({}) bucketized to {}, using {}", oldGenUsedPercent, bucket.toString(), bucketCalculator.toString());
+                if (bucket == UsageBucket.UNDER_UTILIZED || bucket == UsageBucket.HEALTHY_WITH_BUFFER) {
+                  return true;
+                } else {
+                  return false;
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+    return true;
```

Comment:
There are multiple level of nesting here with `for-if-for-if`, can we break down this function into two or more or use lambda ?

Reply:
I have broken the function into two. I will keep thinking about making the loop nicer to look at if I find one. We may not block on that though.